### PR TITLE
memory viewer: Implement RSX mode

### PR DESCRIPTION
**Pull request by elad335**
## Enhancements
* Implement RSX memory manager hook,
* Set the ground for RSX modes of register editor and instruction editor, do not use shared ptrs directly.
* Make register editor and instruction editor modeless to allow to copypaste values from thread context etc in the background. (they close automatically after switching thread)

## Bugfixes
* Fixed crash when viewing an image that was out of bounds of SPU LS.

### CODE REVIEW

**Comment by Nekotekina** [comment]
Maybe it's time to derive rsx::thread from cpu_thread, it doesn't even need to support all functionality

**Comment by elad335** [comment]
Updated pr to use RSX memory viewer mode.

File: rpcs3/rpcs3qt/debugger_frame.cpp [hunk 1]
```diff
@@ -522,78 +570,101 @@ void debugger_frame::UpdateUnitList()
...
-			if (old_cpu == var_cpu) m_choice_units->setCurrentIndex(m_choice_units->count() - 1);
+			on_select(render->id, *render);
 		}
 	}
 
+	// Close dialogs which are tied to the specific thread selected
+	if (!reselected)
+	{
+		if (auto ptr = m_reg_editor) ptr->close();
+		if (auto ptr = m_inst_editor) ptr->close();
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/debugger_frame.cpp:588]: why not like this?
```suggestion
		if (m_reg_editor) m_reg_editor->close();
		if (m_inst_editor) m_inst_editor->close();
```

File: rpcs3/rpcs3qt/instruction_editor_dialog.cpp [hunk 2]
```diff
@@ -13,17 +13,33 @@ constexpr auto qstr = QString::fromStdString;
...
+		{
+			*storage_pointer = nullptr;
+		});
+	}
+
+	const auto cpu = m_get_cpu();
+
+	if (!cpu)
+	{
+		close();
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/instruction_editor_dialog.cpp:38]: this is horrible...
Never close a dialog in the constructor.

File: rpcs3/rpcs3qt/instruction_editor_dialog.cpp [hunk 3]
```diff
@@ -13,17 +13,33 @@ constexpr auto qstr = QString::fromStdString;
...
 	setWindowTitle(tr("Edit instruction"));
 	setAttribute(Qt::WA_DeleteOnClose);
 	setMinimumSize(300, sizeHint().height());
 
-	m_cpu_offset = m_cpu->id_type() == 2 ? static_cast<spu_thread&>(*m_cpu).ls : vm::g_sudo_addr;
+	if (storage_pointer)
+	{
+		connect(this, &QDialog::finished, [storage_pointer](int)
+		{
+			*storage_pointer = nullptr;
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/instruction_editor_dialog.cpp:30]: what are you trying to achieve here ?

Comment (elad335) [reply rpcs3/rpcs3qt/instruction_editor_dialog.cpp:30]: Setting m_kernel_explorer, m_inst_editor and m_reg_editor to nullptr after the dialog is closed.

Comment (Megamouse) [reply rpcs3/rpcs3qt/instruction_editor_dialog.cpp:30]: why

Comment (elad335) [reply rpcs3/rpcs3qt/instruction_editor_dialog.cpp:30]: Because I check if they are nullptr.

Comment (Megamouse) [reply rpcs3/rpcs3qt/instruction_editor_dialog.cpp:30]: then use this and don't overcomplicate things...
connect(m_reg_editor, &QDialog::finished, this, [this]() { m_reg_editor = nullptr; });

File: rpcs3/rpcs3qt/instruction_editor_dialog.cpp (hunk 2 above)
Comment (elad335) [reply rpcs3/rpcs3qt/instruction_editor_dialog.cpp:38]: Qt events are not processed here, the constructor will not call tthe destructor himself. So why not?

Comment (Megamouse) [reply rpcs3/rpcs3qt/instruction_editor_dialog.cpp:38]: Because the object itself gets destroyed and can't be used -> crash

Comment (elad335) [reply rpcs3/rpcs3qt/instruction_editor_dialog.cpp:38]: Made show() not be called after this case, so it's safe.

File: rpcs3/rpcs3qt/instruction_editor_dialog.cpp (hunk 3 above)
Comment (elad335) [reply rpcs3/rpcs3qt/instruction_editor_dialog.cpp]: Done.

File: rpcs3/rpcs3qt/debugger_frame.cpp [hunk 4]
```diff
@@ -500,15 +553,15 @@ void debugger_frame::UpdateUnitList()
...
 		if (emu_state == system_state::stopped) return;
 
-		QVariant var_cpu = QVariant::fromValue<std::weak_ptr<cpu_thread>>(
-			id >> 24 == 1 ? static_cast<std::weak_ptr<cpu_thread>>(idm::get_unlocked<named_thread<ppu_thread>>(id)) : idm::get_unlocked<named_thread<spu_thread>>(id));
-
-		m_choice_units->addItem(qstr(cpu.get_name()), var_cpu);
-		if (old_cpu == var_cpu) m_choice_units->setCurrentIndex(m_choice_units->count() - 1);
+		QVariant var_cpu = QVariant::fromValue<std::pair<cpu_thread*, u32>>(std::make_pair(&cpu, id));
+		m_choice_units->addItem(qstr(id >> 24 == 0x55 ? "RSX[0x55555555]" : cpu.get_name()), var_cpu);
+		if (!reselected && old_cpu == var_cpu) m_choice_units->setCurrentIndex(m_choice_units->count() - 1), reselected = true;
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/debugger_frame.cpp:564]: scope this please. no need to squeeze the code like this

Comment (elad335) [reply rpcs3/rpcs3qt/debugger_frame.cpp]: Fixed.

**Pull request merged**